core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java

```diff
@@ -0,0 +1,810 @@
+    mSyncPathJobs = new LinkedList<>();
+    mMetadataSyncService = concurrencyService;
+  }
+
+  /**
+   * Sync the metadata according the the root path the stream was created with.
+   *
+   * @return true if at least one path was synced
+   */
+  public boolean sync() {
+    // The high-level process for the syncing is:
+    // 1. Given an Alluxio path, determine if it is not consistent with the corresponding UFS path.
+    //     this means the UFS path does not exist, or has metadata which differs from Alluxio
+    // 2. If only the metadata changed, update the inode with the new metadata
+    // 3. If the path does not exist in the UFS, delete the inode in Alluxio
+    // 4. If not deleted, load metadata from the UFS
+    // 5. If a recursive sync, add children inodes to sync queue
+    int syncPathCount = 0;
+    int stopNum = -1; // stop syncing when we've processed this many paths. -1 for infinite
+
+    try {
+      syncInodeMetadata(mRootPath);
+      syncPathCount++;
+      if (mDescendantType == DescendantType.ONE) {
+        // If descendantType is ONE, then we shouldn't process any more paths except for those
+        // currently in the queue
+        stopNum = mSyncMetadataQ.size();
+      }
+
+      // process the sync result for the original path
+      try {
+        mRootPath.traverse();
+      } catch (InvalidPathException e) {
+        throw new RuntimeException(e);
+      }
+    } catch (AccessControlException | BlockInfoException | FileAlreadyCompletedException
+        | FileDoesNotExistException | InvalidFileSizeException | InvalidPathException
+        | IOException e) {
+      LOG.warn("FAILED TO SYNC METADATA: {}", e.getMessage(), e);
+    } finally {
+      // regardless of the outcome, remove the UfsStatus for this path from the cache
+      mStatusCache.remove(mRootPath.getUri());
+      // downgrade so that if operations are parallelized, the lock on the root doesn't restrict
+      // concurrent operations
+      mRootPath.downgradeToPattern(LockPattern.READ);
+    }
+
+    // Process any children after the root.
+    while (!mSyncMetadataQ.isEmpty() || !mSyncPathJobs.isEmpty()) {
+      if (Thread.currentThread().isInterrupted()) {
+        LOG.warn("Metadata syncing was interrupted before completion");
+        break;
+      }
+      // There are still paths to process
+      // First, remove any futures which have completed. Add to the sync path count if they sync'd
+      // successfully
+      while (true) {
+        Future<Boolean> job = mSyncPathJobs.peek();
+        if (job == null || !job.isDone()) {
+          break;
+        }
+        // remove the job because we know it is done.
+        if (mSyncPathJobs.poll() != job) {
+          throw new IllegalStateException("Last node to be de-queued was not equal to the expected"
+              + "head of queue");
+        }
+        try {
+          // we synced the path successfully
+          if (job.get()) {
+            syncPathCount++;
```

Comment:
if some of the tasks failed for some reason, will the count accounting still work? If I want to load a directory with 10 children, but 1 failed, will it still stop after all 10 finished (fail or succeed)?

Reply:
Technically for more complicated directory structures it depends, (directory high up can fail, causing its children not to sync) but generally the accounting is only for paths which successfully synced. So if all paths fail, then this function returns false. If at least one path syncs, we return true.

We can keep track of the number of failed paths too and report them as debug logs?